src/client/components/AddDirectMessageModal.tsx

```diff
@@ -0,0 +1,242 @@
+import type { Dispatch, SetStateAction } from 'react';
+import React, {
+  useCallback,
+  useContext,
+  useEffect,
+  useMemo,
+  useState,
+} from 'react';
+import { styled } from 'styled-components';
+import { useNavigate } from 'react-router-dom';
+import { user, Avatar as DefaultAvatar } from '@cord-sdk/react';
+import { ChatBubbleLeftRightIcon } from '@heroicons/react/20/solid';
+import { XIcon } from 'src/client/components/Buttons';
+import { Modal as DefaultModal } from 'src/client/components/Modal';
+import { useLazyAPIFetch } from 'src/client/hooks/useAPIFetch';
+import { ChannelsContext } from 'src/client/context/ChannelsContext';
+import { EVERYONE_ORG_ID } from 'src/client/consts/consts';
+import { Colors } from 'src/client/consts/Colors';
+import { Name } from 'src/client/components/Name';
+import { DM_CHANNEL_PREFIX } from 'src/common/consts';
+
+export function AddDirectMessageModal({
+  isOpen,
+  setModalOpen,
+}: {
+  isOpen: boolean;
+  setModalOpen: Dispatch<SetStateAction<boolean>>;
+}) {
+  const {
+    orgMembers: allOrgMembers,
+    loading,
+    hasMore,
+    fetchMore,
+  } = user.useOrgMembers({
+    organizationID: EVERYONE_ORG_ID,
+  });
+  const viewer = user.useViewerData();
+
+  useEffect(() => {
+    if (!loading && hasMore) {
+      void fetchMore(50);
+    }
+  }, [hasMore, loading, fetchMore]);
+
+  const addableUsers = useMemo(() => {
+    return allOrgMembers
+      .filter((om) => om.id !== viewer?.id)
+      .sort((a, b) => a.displayName.localeCompare(b.displayName));
+  }, [allOrgMembers, viewer?.id]);
+
+  const [usersToAdd, setUsersToAdd] = useState<string[]>([]);
+
+  const { refetch: refetchChannels } = useContext(ChannelsContext);
+  const update = useLazyAPIFetch();
+  const navigate = useNavigate();
+
+  const createDM = useCallback(() => {
+    if (viewer === undefined || usersToAdd.length === 0) {
+      return;
+    }
+    const channelUsers = [viewer.id, ...usersToAdd];
+    const channelName = DM_CHANNEL_PREFIX + channelUsers.sort().join(',');
+
+    void update(`/channels/${encodeURIComponent(channelName)}`, 'PUT', {
```

Comment:
Should this be a special endpoint on the server, with the normal endpoint prohibiting creating DMs? I think as-is some weird things will happen if I try to create a DM channel between two people who aren't me. And we can also move the sorting logic (line 62) onto the server too.

I'm slightly less worried about deliberate exploitation as I am about bugs or other data model issues (e.g., someone trying to debug something and doing the above by accident).

Reply:
This diff also changes the server-side code to handle this case explicitly (both rejecting any attempt to create a DM channel without the viewer in it and re-doing the sorting).  I figured it was easier to have only one endpoint that creates new channels than have two, but could split it out if you feel strongly.

Reply:
Ah missed that. Slight pref for two endpoints, since the logic is pretty split anyway, but I don't feel strongly at all.